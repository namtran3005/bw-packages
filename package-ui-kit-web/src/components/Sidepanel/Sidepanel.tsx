import * as React from 'react';

import styled from 'styled-components';
import { Colors, Spacing, ColorsType } from '@bitwala-cryptobank-squad/package-theme';

import { GridLayoutProps } from '../..';
import { Logo, LogoProps } from '../Logo';

interface SidepanelProps extends GridLayoutProps {
  backgroundImage?: string;
  background?: ColorsType;
  borderless?: boolean;
  fixed?: boolean;
  zIndex?: number;
  logoProps?: LogoProps;
  logo?: boolean;
  children?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
  inLoggedOutLayout?: boolean;
}

const defaultBackgroundColor = 'backgroundLight';

const SidepanelContentWrapper = styled.div.withConfig({
  shouldForwardProp: (prop) =>
    !['backgroundImage', 'background', 'fixed', 'zIndex'].includes(prop),
})`
  background: red;
`;

const SidepanelComponent = styled.div.withConfig({
  shouldForwardProp: (prop) =>
    ![
      'backgroundImage',
      'background',
      'fixed',
      'zIndex',
      'columnsTemplate',
      'inLoggedOutLayout',
      'logoPosition',
    ].includes(prop),
})<SidepanelProps>`
  height: 100vh;
  max-height: 100vh;
  position: sticky;
  left: 0;
  bottom: 0;
  top: 0;
  right: 0;
  display: block;

  ${SidepanelContentWrapper} {
    height: 100vh;
    max-height: 100vh;
    ${({ borderless }) =>
      !borderless &&
      `border-right: solid 1px ${Colors.getWithAlpha('text', 0.15)};`}
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    background: ${({ background }) =>
      Colors.get(background ? background : defaultBackgroundColor)};

    &:before {
      content: '';
      background-image: ${(props) =>
        props.backgroundImage ? `url(${props.backgroundImage})` : 'none'};
      background-size: cover;
      background-position: center;
      opacity: 1;
      top: 0;
      left: 0;
      bottom: 0;
      right: 0;
      position: absolute;
      z-index: 0;
    }
  }
`;

const LogoWrapperComponent: React.FunctionComponent<LogoProps> = (props) => (
  <Logo {...props} />
);

const LogoComponent = styled(LogoWrapperComponent)`
  position: relative;
  margin: ${Spacing.unit * 4}px ${Spacing.unit * 3}px;
  z-index: 10;
`;

function Sidepanel({
  children,
  logo,
  logoProps = { variant: 'logomark' },
  ...cleanProps
}: SidepanelProps): React.ReactElement {
  return (
    <SidepanelComponent {...cleanProps}>
      <SidepanelContentWrapper>
        {logo && (
          <div
            {...(cleanProps.inLoggedOutLayout && {
              style: { position: 'absolute', top: '100px' },
            })}
          >
            <LogoComponent {...logoProps} />
          </div>
        )}
        {children}
      </SidepanelContentWrapper>
    </SidepanelComponent>
  );
}

export { Sidepanel, SidepanelProps };
