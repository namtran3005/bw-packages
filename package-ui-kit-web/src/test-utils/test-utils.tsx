import {
  render as rtlRender,
  waitFor,
  RenderOptions as RTLRenderOptions,
} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { WebAppThemeProvider } from '../theme';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';

type RenderOptions = RTLRenderOptions;

function renderWithTestHarness(
  ui: React.ReactElement,
  renderOptions: RenderOptions = {}
) {
  function Wrapper({ children }: any) {
    return (
      <MemoryRouter>
        <WebAppThemeProvider>{children}</WebAppThemeProvider>
      </MemoryRouter>
    );
  }

  return rtlRender(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });
}

// Use if you're in a hurry (better than nothing)
// be sure to verify no memory leak from implicit
// waitFor by adding these flags to jest: --runInBand --detectOpenHandles
async function renderAct(ui: React.ReactElement, options: RenderOptions = {}) {
  const view = renderWithTestHarness(ui, options);
  return await waitFor(() => view, {
    timeout: 5000,
    onTimeout: (e) => {
      // eslint-disable-next-line no-console
      console.error(e);
      return e;
    },
  });
}

export * from '@testing-library/react';

export { renderWithTestHarness, renderAct, userEvent };
