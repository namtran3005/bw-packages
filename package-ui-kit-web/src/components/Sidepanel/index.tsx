import React from 'react';

import {
  SidepanelContent as SidepanelContentComponent,
  SidepanelContentProps,
} from '../SidepanelContent';
import {
  SidepanelFooter as SidepanelFooterComponent,
  SidepanelFooterProps,
} from '../SidepanelFooter';
import { Sidepanel as SidepanelComponent, SidepanelProps } from './Sidepanel';

function Sidepanel(props: Partial<SidepanelProps>): React.ReactElement {
  return <SidepanelComponent {...props} />;
}

function SidepanelContent(props: SidepanelContentProps): React.ReactElement {
  return <SidepanelContentComponent {...props} />;
}

function SidepanelFooter(props: SidepanelFooterProps): React.ReactElement {
  return <SidepanelFooterComponent {...props} />;
}

Sidepanel.prototype.Content = SidepanelContent;
Sidepanel.Content = Sidepanel.prototype.Content;
Sidepanel.prototype.Footer = SidepanelFooter;
Sidepanel.Footer = Sidepanel.prototype.Footer;

export { Sidepanel, SidepanelProps };
