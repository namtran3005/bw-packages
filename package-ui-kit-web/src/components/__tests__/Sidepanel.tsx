import * as React from 'react';

import { renderWithTestHarness } from '../../test-utils/test-utils';
import { Sidepanel } from '../Sidepanel';

describe('Sidepanel', () => {
  it('should render Sidepanel', () => {
    const { baseElement } = renderWithTestHarness(<Sidepanel />);
    expect(baseElement).toMatchSnapshot();
  });
  it('should render Sidepanel with logo - 1', () => {
    const { baseElement } = renderWithTestHarness(<Sidepanel logo />);
    expect(baseElement).toMatchSnapshot();
  });
  it('should render Sidepanel with logo - 2', () => {
    const { baseElement } = renderWithTestHarness(
      <Sidepanel backgroundImage="url" />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
