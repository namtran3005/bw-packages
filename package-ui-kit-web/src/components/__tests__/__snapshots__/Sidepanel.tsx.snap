// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidepanel should render Sidepanel 1`] = `
.c2 {
  background: red;
}

.c0 {
  height: 100vh;
  max-height: 100vh;
  position: -webkit-sticky;
  position: sticky;
  left: 0;
  bottom: 0;
  top: 0;
  right: 0;
  display: block;
}

.c0 .c1 {
  height: 100vh;
  max-height: 100vh;
  border-right: solid 1px rgba(44,35,46,0.15);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background: #F0F0F0;
}

.c0 .c1:before {
  content: '';
  background-image: none;
  background-size: cover;
  background-position: center;
  opacity: 1;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  position: absolute;
  z-index: 0;
}

<body>
  <div>
    <div
      class="c0"
    >
      <div
        class="c1 c2"
      />
    </div>
  </div>
</body>
`;

exports[`Sidepanel should render Sidepanel with logo - 1 1`] = `
<body>
  .c3.c3.c3 {
  position: relative;
  display: inline-block;
  cursor: initial;
}

.c2 {
  background: red;
}

.c0 {
  height: 100vh;
  max-height: 100vh;
  position: -webkit-sticky;
  position: sticky;
  left: 0;
  bottom: 0;
  top: 0;
  right: 0;
  display: block;
}

.c0 .c1 {
  height: 100vh;
  max-height: 100vh;
  border-right: solid 1px rgba(44,35,46,0.15);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background: #F0F0F0;
}

.c0 .c1:before {
  content: '';
  background-image: none;
  background-size: cover;
  background-position: center;
  opacity: 1;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  position: absolute;
  z-index: 0;
}

.c4 {
  position: relative;
  margin: 32px 24px;
  z-index: 10;
}

<div>
    <div
      class="c0"
    >
      <div
        class="c1 c2"
      >
        <div>
          <span
            class="c3 c4"
          >
            <svg
              fill="none"
              height="48"
              viewBox="0 0 40 40"
              width="48"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M31.016 0H8.984C4.02227 0 0 4.02227 0 8.984V31.016C0 35.9777 4.02227 40 8.984 40H31.016C35.9777 40 40 35.9777 40 31.016V8.984C40 4.02227 35.9777 0 31.016 0Z"
                fill="#2C232E"
              />
              <path
                d="M17.3018 25.2538C17.3018 23.7098 15.9578 22.5645 14.4138 22.5645H9.03516V31.6311H10.2805V27.9418H13.2378L15.8085 31.6271H17.3018L14.6458 27.9324C16.0898 27.8258 17.3018 26.7191 17.3018 25.2538ZM14.4138 26.7991H10.2805V23.7098H14.4138C15.2605 23.7098 16.0578 24.3071 16.0578 25.2538C16.0578 26.2005 15.2605 26.7991 14.4138 26.7991Z"
                fill="#D4D2D2"
              />
              <path
                d="M22.7305 23.7098H26.2158V30.4818H22.7305V31.6271H30.9971V30.4818H27.5105V23.7098H30.9971V22.5645H22.7305V23.7098Z"
                fill="#D4D2D2"
              />
              <path
                d="M16.0578 16.2884H15.7085L11.3765 8.37109H9.03516V17.4338H10.2805V9.51643H10.6298L14.9618 17.4338H17.3018V8.37109H16.0578V16.2884Z"
                fill="#D4D2D2"
              />
              <path
                d="M29.7518 8.37109V13.6498C29.7518 15.3924 28.6065 16.6871 26.8638 16.6871C25.1211 16.6871 23.9758 15.3924 23.9758 13.6498V8.37109H22.7305V13.6498C22.7305 16.0391 24.4731 17.8324 26.8638 17.8324C29.2545 17.8324 30.9971 16.0391 30.9971 13.6498V8.37109H29.7518Z"
                fill="#D4D2D2"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Sidepanel should render Sidepanel with logo - 2 1`] = `
<body>
  .c2 {
  background: red;
}

.c0 {
  height: 100vh;
  max-height: 100vh;
  position: -webkit-sticky;
  position: sticky;
  left: 0;
  bottom: 0;
  top: 0;
  right: 0;
  display: block;
}

.c0 .c1 {
  height: 100vh;
  max-height: 100vh;
  border-right: solid 1px rgba(44,35,46,0.15);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background: #F0F0F0;
}

.c0 .c1:before {
  content: '';
  background-image: url(url);
  background-size: cover;
  background-position: center;
  opacity: 1;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  position: absolute;
  z-index: 0;
}

<div>
    <div
      class="c0"
    >
      <div
        class="c1 c2"
      />
    </div>
  </div>
</body>
`;
